src/Compilers/Core/Portable/SourceGeneration/Nodes/ISyntaxInputNode.cs

```diff
@@ -16,7 +17,7 @@ internal interface ISyntaxInputBuilder
     {
         ISyntaxInputNode SyntaxInputNode { get; }
 
-        void VisitTree(SyntaxNode root, EntryState state, SemanticModel? model, CancellationToken cancellationToken);
+        void VisitTree(Lazy<SyntaxNode> root, EntryState state, SemanticModel? model, CancellationToken cancellationToken);
```

Comment:
❔ Can we not make this accept `SyntaxTree` instead? The call to `GetRoot` is cheap at the point where it's required.

Reply:
Is it cheap to call it multiple times? If you have multiple generators, each one will call `tree.GetRoot()` I didn't want to asssume that subsequent calls would be cheaper...

Reply:
It's unlikely for the call to `GetRoot` to be expensive for the second and subsequent calls (when needed). There is a brief moment where the root could be collected by the GC when moving from one source generator to the next. It could be prevented like this:

```suggestion
        void VisitTree(SyntaxTree tree, ref SyntaxNode? lazyRoot, EntryState state, SemanticModel? model, CancellationToken cancellationToken);
```

Then where you need the tree:

```csharp
var root = LazyInitialization.EnsureInitialized(
    ref lazyRoot,
    static arg => arg.tree.GetRoot(arg.cancellationToken),
    (tree, cancellationToken));
```

Reply:
Hmm, Jared pointed out that if we only get the root when the tree is different, then it's unlikely to be a recoverable tree in the first place, in which case `GetRoot` should be cheap. @sharwell does that seem correct?

Reply:
I'm not sure why that would make a difference.